homeassistant/components/shopping_list.py

```diff
@@ -175,23 +215,34 @@ def async_update(self, item_id, info):
         return item
 
     @callback
-    def async_clear_completed(self):
+    def async_clear_completed(self, list_id):
         """Clear completed items."""
-        self.items = [itm for itm in self.items if not itm['complete']]
-        self.hass.async_add_job(self.save)
+        lis = next((li for li in self.lists if li['id'] == list_id), None)
 
-    @asyncio.coroutine
-    def async_load(self):
-        """Load items."""
-        def load():
-            """Load the items synchronously."""
-            return load_json(self.hass.config.path(PERSISTENCE), default=[])
+        if lis is None:
+            raise KeyError
 
-        self.items = yield from self.hass.async_add_job(load)
+        lis['items'] = [itm for itm in lis['items'] if not itm['complete']]
+        self.hass.async_add_job(self.save)
 
-    def save(self):
+    async def async_load(self):
+        """Load lists."""
+        self.lists = await self.hass.helpers.storage.async_migrator(
```

Comment:
Sorry,  I don't follow on this one

Reply:
@balloob are you able to clarify this at all for me? 🙏

Reply:
To avoid confusion, let me use the word `array` instead of Python's data structure `list`.

In the storage component, you store an array of lists. You keep the same data structure when loaded into memory. This means that for you to look up a shopping list, you're doing a `next(…)` call to extract the list from the array.

Instead, in memory, you want to convert the array of lists to a dictionary, keyed by the ID of the list. That way you're able to do fast look ups. You will however need to preserve the order the lists are created, as that might be significant. Standard dictionaries are only ordered starting Python 3.6, so you need to use `OrderedDict`.

```python
from collections import OrderedDict

data = await self.hass.helpers.storage.async_migrator(…)
lists = OrderedDict()
if data is None:
    …

for list in data:
    lists[list['id']] = list

self.lists = lists
```

Then update `_data_to_save` to convert the dictionary to a list again: `list(self.lists.values())`